src/runtime_src/core/common/drv/xrt_cu.c

```diff
@@ -9,107 +9,147 @@
-			xrt_cu_start(xcu);
-			/* xcmd should always point to next waiting
-			 * to submit command
-			 */
-			if (xcmd != last_xcmd)
-				xcmd = list_next_entry(xcmd, list);
-			else
-				xcmd = NULL;
-		} else {
-			/* Run out of credit and still have xcmd in rq.
-			 * In this case, only do wait one more command done.
-			 */
-			process_sq_once(xcu);
-		}
-	}
 
-	/* Some commands maybe not completed
-	 * or they are completed but haven't beed processed
-	 * Do not wait, get pending command first.
-	 */
-	if (!list_empty(&xcu->rq))
-		list_splice_tail_init(&xcu->rq, &xcu->sq);
+	if (!xcu->num_rq)
+		return 0;
+
+	xcmd = list_first_entry(&xcu->rq, struct kds_command, list);
+
+	if (!xrt_cu_get_credit(xcu))
+		return 0;
+
+	/* if successfully get credit, you must start cu */
+	xrt_cu_config(xcu, (u32 *)xcmd->info, xcmd->isize, 0);
+	xrt_cu_start(xcu);
+
+	/* Move xcmd to submmited queue */
+	list_move_tail(&xcmd->list, &xcu->sq);
+	--xcu->num_rq;
+	++xcu->num_sq;
+
+	return 1;
 }
 
 int xrt_cu_thread(void *data)
 {
 	struct xrt_cu *xcu = (struct xrt_cu *)data;
 	unsigned long flags;
-
-	while (1) {
-		// Check num_pq here
+	int ret = 0;
+
+	while (!xcu->stop) {
+		/* process run queue would return 1 if submit a command to CU
+		 * Let's try submit another command
+		 */
+		if (process_rq(xcu))
+			continue;
+		/* process completed queue before submitted queue, for two reasons
+		 * - The last submitted command may be still running
+		 * - while handling completed queue, running command might done
+		 * - process_sq will check CU status, which is thru slow bus
+		 */
+		process_cq(xcu);
+		process_sq(xcu);
+
+		/* Continue until run queue emtpy */
+		if (xcu->num_rq)
+			continue;
+
+		if (!xcu->num_sq && !xcu->num_cq)
+			if (down_interruptible(&xcu->sem))
+				ret = -ERESTARTSYS;
+
+		/* Get pending queue command number without lock.
+		 * The idea is to reduce the possibility of conflict on lock.
+		 * Need to check pending command number again after lock.
+		 */
+		if (!xcu->num_pq)
+			continue;
 		spin_lock_irqsave(&xcu->pq_lock, flags);
-		// double check
-		if (xcu->num_pq > 0) {
+		if (xcu->num_pq) {
 			list_splice_tail_init(&xcu->pq, &xcu->rq);
+			xcu->num_rq = xcu->num_pq;
 			xcu->num_pq = 0;
 		}
 		spin_unlock_irqrestore(&xcu->pq_lock, flags);
-
-		/* !!!Do not change the priority! */
-		if (!list_empty(&xcu->rq)) {
-			process_rq(xcu);
-		} else if (!list_empty(&xcu->sq)) {
-			process_sq_once(xcu);
-		} else {
-			while (down_timeout(&xcu->sem, 1000) == -ETIME) {
-				if (kthread_should_stop())
-					return 0;
-			}
-		}
 	}
 
-	return 0;
+	return ret;
 }
 
 void xrt_cu_submit(struct xrt_cu *xcu, struct kds_command *xcmd)
 {
 	unsigned long flags;
 
+	/* Add command to pending queue
+	 * wakeup CU thread if it is the first command
+	 */
 	spin_lock_irqsave(&xcu->pq_lock, flags);
 	list_add_tail(&xcmd->list, &xcu->pq);
-	if (xcu->num_pq == 0)
-		up(&xcu->sem);
 	++xcu->num_pq;
+	if (xcu->num_pq == 1)
+		up(&xcu->sem);
```

Comment:
you could cache the result of "xcu->num_pq == 1" and do the up after releasing the lock.

Reply:
done